### PR TITLE
feat(bridge): downstream workspace settings propagation

### PR DESCRIPTION
**Pull request by atusy**
## Summary

Make the bridge own and serve each downstream language server's **workspace configuration**, so a downstream server (rust-analyzer, lua-ls, …) receives its settings both at startup and on runtime change — not only once via `initializationOptions`.

The editor configures *kakehashi*, not `rust-analyzer`, so it cannot answer a downstream server's `workspace/configuration` request. The bridge therefore serves the configuration itself, sourced from its already-merged settings tree.

ADR: `docs/architecture-decisions/downstream-settings-propagation.md`.

## What changed

- **`settings` config field** — new `languageServers.<name>.settings` (`BridgeServerConfig.settings`), opaque passthrough, deep-merged across config layers like `initializationOptions`. Distinct from `initializationOptions` (consumed only at `initialize`); `settings` flows over `didChangeConfiguration` and answers `workspace/configuration`.
- **(b) Answer pulls** — new `workspace/configuration` handler resolves each item's dotted `section` against this server's settings root and returns an `LSPAny[]` of the same length/order as `items`, JSON `null` for misses. Backed by a per-connection settings cell shared with the reader.
- **(a) Advertise + seed** — advertise `workspace.configuration` downstream, **gated per-server on `settings.is_some()`** so a server configured only via `initializationOptions` (every existing setup) is not flipped to pulling and answered `null`. After `initialized`, push `workspace/didChangeConfiguration` (reading the live cell, queued before the `Ready` flip so it precedes any `didOpen`).
- **(c) Re-propagate** — on any merged-settings change (`apply_shared_settings` choke point: `didChangeConfiguration`, auto-install reload, initialize), re-resolve each live connection's settings, diff against its cell, and on change re-store the cell + push best-effort to `Ready` connections (initializing ones get the cell advanced and are pushed by (a) after `initialized`). Returns the push count; unchanged servers get nothing.

## Scope / deferred

- Upstream pull (kakehashi → editor `workspace/configuration`) — only needed for pull-model editors; deferred.
- `scopeUri` / multi-root resolution — ignored; one per-server value answers all scopes.
- Guaranteed push delivery — a dropped push self-heals on the next pull or respawn (single-cell, no retry baseline). The narrow (a)/(c) handshake-window race is the accepted, ADR-documented epoch-class race.

## Review

Converged through `/review` (10-perspective), codex, and pi-review before opening:
- per-server capability gating, live-cell post-init push, push-count observability;
- LSP init-ordering (no `didChangeConfiguration` to an initializing server; push queued before `Ready`);
- section-resolution edge cases (empty segments, null leaf, array mid-traversal).

## Testing

- `cargo test --lib` green (2219 passed); clippy `--lib` clean.
- Unit tests cover section resolution + response shape, `settings` deep-merge, the notification builder, the shared settings cell, `propagate_settings` selective update / initializing-skip / empty-pool, and capability gating.
- e2e deferred to CI.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/lsp/bridge/pool.rs [hunk 1]
```diff
@@ -460,6 +461,55 @@ impl LanguageServerPool {
...
+            // Only notify a Ready connection: sending
+            // `workspace/didChangeConfiguration` to a still-initializing server
+            // would precede its `initialized` and violate LSP ordering. An
+            // Initializing connection instead gets its (now-updated) cell pushed
+            // by path (a) once the handshake completes. Count only a successful
+            // enqueue, so the returned count never reports a dropped push.
+            if handle.state() == ConnectionState::Ready {
+                let result = handle.send_notification(build_did_change_configuration_notification(
+                    resolved.unwrap_or(serde_json::Value::Null),
+                ));
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Addressed the clone optimizations (Arc-first store, inner Value cloned only for a Ready push). The `Arc::ptr_eq`-against-the-map liveness recheck is not needed here: `propagate_settings` holds the `connections` lock for the whole loop and iterates the map's live values directly (never retrieve-then-release), so a respawn — which takes the same lock — cannot replace a handle mid-iteration. Added a comment noting this.

[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Addressed the clone optimizations (Arc-first store, inner Value cloned only for a Ready push). The `Arc::ptr_eq`-against-the-map liveness recheck is not needed here: `propagate_settings` holds the `connections` lock for the whole loop and iterates the map's live values directly (never retrieve-then-release), so a respawn — which takes the same lock — cannot replace a handle mid-iteration. Added a comment noting this.

File: src/lsp/bridge/pool.rs [hunk 2]
```diff
@@ -1473,6 +1539,22 @@ impl LanguageServerPool {
...
+                    // processes a document under default config. Read the *live*
+                    // cell, not the spawn-time value, so a path-c re-propagation
+                    // that landed during the handshake is reflected and the push
+                    // agrees with what (b) would answer. Best-effort: a dropped
+                    // notification self-heals on the next pull.
+                    if let Some(settings) = handle_for_handshake.current_settings() {
+                        handle_for_handshake.send_notification(
+                            build_did_change_configuration_notification((*settings).clone()),
+                        );
+                    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Good catch — fixed. The post-`initialized` push now fires when the server advertised `configuration` OR the cell holds settings, sending JSON `null` when the cell is empty, so a clear during the handshake reaches push-model servers instead of leaving them on stale settings.

File: src/lsp/bridge/workspace/configuration.rs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    match section {
+        None | Some("") => root.clone(),
+        Some(path) => {
+            let mut current = root;
+            for segment in path.split('.') {
+                match current.get(segment) {
+                    Some(next) => current = next,
+                    None => return Value::Null,
+                }
+            }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/workspace/configuration.rs:106]: Fixed — `resolve_section` now rejects empty path segments explicitly, so a malformed section resolves to null regardless of whether settings contains a literal empty-string key. Added a test with a `""`-key trap.
